ext/drbbdup/drbbdup.c

```diff
@@ -1590,7 +1590,8 @@ drbbdup_instrument_dups(void *drcontext, void *tag, instrlist_t *bb, instr_t *in
             }
         }
         drreg_restore_all(drcontext, bb, instr);
-    } else if (drbbdup_is_at_end(instr) && !is_last_special) {
+    } else if (drbbdup_is_at_end(instr) &&
+               (!is_last_special || drbbdup_ilist_has_unending_emulation(bb))) {
```

Comment:
Trying to understand this: so `is_at_end_initial` is always true for the end of all block copies, or at least it seems like that from its header comment.

But you said:

> Normal case: drbbdup_is_at_end_initial returns true, so drbbdup_instrument_instr and drreg_restore_all are called
> Basic block just before the rep block: drbbdup_is_at_end_initial is false, drreg_restore_all is called from the !is_last_special case and drbbdup_instrument_instr is called from the final else
> For the rep block: we don't go into the !is_last_special case, and we don't reach the final else because drbbdup_is_exit_jmp_emulation_marker is true

Does that mean this comment for `drbbdup_is_at_end_initial` is wrong?

> If there are no emulation labels this is equivalent to drbbdup_is_at_end().

I'm finding this very confusing: why does the original code on line 1593 have a check `!is_last_special`?

If you've got an understanding of it, maybe you could add and augment comments here and in the headers for these query functions?

Reply:
Yes, that comment looks incorrect to me. You can see that it returns false if drmgr_is_emulation_start returns false https://github.com/DynamoRIO/dynamorio/blob/f59f1b3a2183ba837177d1cac62ce7b706ed555e/ext/drbbdup/drbbdup.c#L667-L685

I think the check exists for handling the scenario that you brought up

> You're saying the drbbdup identification of the end of a copied block relies on an emulation start label, which is not there for rep-str. Wouldn't it also not be there for a block that doesn't end in a branch/syscall? Does such a block also have this problem?

Reply:
So is the `!is_last_special` trying to detect whether control already went into the case above because `drbbdup_is_at_end_initial` returned true?  It's equating the presence of an emulation start label with a special instr being there (which requires the emulation b/c drbbdup can't put the real instr in each copy).  But if the rep expansion does have `is_last_special` why doesn't it have an emulation label for it?  Looking around: I see this explanation:
https://github.com/DynamoRIO/dynamorio/blob/f59f1b3a2183ba837177d1cac62ce7b706ed555e/ext/drbbdup/drbbdup.c#L474-L475

So it seems that the real intent of the `drbbdup_is_at_end(instr) && xxx` is to capture all cases that do not go into the `drbbdup_is_at_end_initial`, but not to trigger for any cases that do, right?  So I would suggest to not add further conditions to `xxx` which may just make everything more fragile, but to instead have a local boolean in this function which says "inserted_restore_all" and set that in the `*end_initial` case and replace `xxx` with `!inserted_restore_all`.  And please add comments to help a future reader understand all this.

Reply:
So we only want to handle those specific cases, and other cases are handled by the next else condition

https://github.com/DynamoRIO/dynamorio/blob/f59f1b3a2183ba837177d1cac62ce7b706ed555e/ext/drbbdup/drbbdup.c#L1593-L1597

Reply:
The insertion of drreg_restore_all() is also ignoring/hiding the instruction by not calling drbbdup_instrument_instr so there is no need to hit that next else which says "Ignore": these earlier conditions are also ignoring if they don't hand the instr to drbbdup_instrument_instr.  Could combine all the drbbdup_is_at_end() into one condition with sub-conditions if that seems clearer.

Reply:
Resolving this since we are discussing this below.